### PR TITLE
Add middle click copy

### PR DESCRIPTION
**Pull request by leovilok**
I think it's the most expected middle mouse button behaviour.

### CODE REVIEW

**Comment by jarun** [comment]
Why? We already have right click copy in.

**Comment by jarun** [comment]
We already have right click add to to selection in. Why do we have to have middle click copy and not move? Are there any other file managers which supports this? Even if someone mis-clicks a copy would be issued. Why would we want this?

**Comment by jarun** [comment]
How about something much more useful? Like cycle context color?

**Comment by leovilok** [comment]
> We already have right click add to to selection in. Why do we have to have middle click copy and not move?
The idea is to be able to copy files around using only the mouse. Moving the files could work too.

> Are there any other file managers which support this?
I don't think so. I think that graphical file managers tend to open a new tab on middle click, just as browsers. But middle click is used to paste (text) selection everywhere on x11.

> Even if someone mis-clicks a copy would be issued. Why would we want this?
This is indeed an issue, as most misclicks. It behaves just as if you pressed 'p'. I don't know if middle mouse miss click happen that often, but it might be more surprising than a keyboard missclick.

> How about something much more useful? Like cycle context color?
This deserves its own (keyboard) key first.

But middle click could just set `presel` to a custom, configurable, key. So that anyone can compile in the behaviour of any other allready available command.

What do you think about that ?
What would be the default behaviour ?

**Comment by jarun** [comment]
> I don't know if middle mouse miss click happen that often

It happens with me often while scrolling the wheel.

> This deserves its own (keyboard) key first.

It doesn't have and that's something that would make it unique. But I do have an idea here. We already detect clicking on the context number. When left-clicked on the current context number, we can change the color to the next color code. We can have the same behaviour if we press the same context number which you are in.

But would that be confusing? What do you say?

> But middle click could just set presel to a custom, configurable, key.

That's a nice idea. By default we can use an unassigned key. But how would someone assign a Ctrl-key?

**Comment by leovilok** [comment]
> When left-clicked on the current context number, we can change the color to the next color code. We can have the same behaviour if we press the same context number which you are in.
> 
> But would that be confusing? What do you say?
Might be a little confusing and hard to discover, but it's harmless, so I don't think it would be an issue.

> That's a nice idea. By default we can use an unassigned key. But how would someone assign a Ctrl-key?

If it's in `config.h` we can just use the `CONTROL()` macro.
If it's an env variable it can be passed raw in the shell or in an editor, usually with Ctrl+V.

**Comment by leovilok** [comment]
Middle click action is now configurable with the `NNN_MIDDLECLICK` env variable.

**Comment by jarun** [comment]
Can you share a working config example for middle click for me to test?

**Comment by jarun** [comment]
Say for `^R`.

File: nnn.1 [hunk 1]
```diff
@@ -375,6 +375,16 @@ separated by \fI;\fR:
     export NNN_TRASH=1
 .Ed
 .Pp
+\fBNNN_MIDDLECLICK:\fR key emulated by a middle mouse click.
+.Bd -literal
+    export NNN_MIDDLECLICK='p'
+
+    NOTES:
+    1. Only the first character is taken into account.
+    2. Use your shell/editor's quoted insert function (usually \fICtrl+V\fR)
```
Comment (jarun) [inline nnn.1]: We need to put an example. I am not able to figure out what we need to set.

File: src/nnn.c [hunk 2]
```diff
@@ -567,8 +570,9 @@ static const char * const messages[] = {
 #define NNN_COLORS 4
 #define NNNLVL 5
 #define NNN_PIPE 6
-#define NNN_ARCHIVE 7 /* strings end here */
-#define NNN_TRASH 8 /* flags begin here */
+#define NNN_MIDDLECLICK 7
```
Comment (jarun) [inline src/nnn.c]: Please use `NNN_MCLICK` for a short and sweet one. We will document it.

**Comment by leovilok** [comment]
> Say for `^R`.

Github won't let me attach the file, probably because of the `^R`, but here it is:
http://ix.io/2dVF
Download it, `source` it, and lanch `./nnn`.

You can do it in the shell too, just type ^V before inserting ^R.

**Comment by jarun** [comment]
Thanks! I'll test this. There are some minor review comments. Please address.

File: nnn.1 (hunk 1 above)
Comment (leovilok) [reply nnn.1]: Well it looks like

```
export NNN_MIDDLECLICK='^R'
```

But if you input it right, ^R is a actually a single character.

Of course if you use `bash` or `zsh` you can write `export NNN_MIDDLECLICK=$'\01'` for `^A`, `export NNN_MIDDLECLICK=$'\02'` for `^B`, etc.
`^R` is `\022`

**Comment by jarun** [comment]
I think we can do better than this. We can take the standard notation in, like:

    export NNN_MCLICK='^V'
and parse the value to set the key to `'V' ^ 0x40`.

**Comment by leovilok** [comment]
> I think we can do better than this. We can take the standard notation in, like:
> 
> ```
> export NNN_MCLICK='^V'
> ```
> 
> and parse the value to set the key to `'V' ^ 0x40`.

Yes.

File: nnn.1 (hunk 1 above)
Comment (jarun) [reply nnn.1]: As discussed, we will use the standard notation and parse. Please add 2 examples in the man page, one for `p` (as it is now) another for `^R`.

Comment (jarun) [reply nnn.1]: Also, please remove the notes.

Comment (leovilok) [reply nnn.1]: Ah, I didn't see that.

**Comment by jarun** [comment]
Thank you!

**Comment by leovilok** [comment]
Current `CONTROL()` macro only works for uppercase letters, it's basically `c ^ 0x40`.
If I replace it with `c & 0x1f` it works for both uppercase and lowercase letters.

**Comment by jarun** [comment]
Thanks! I'll add it along with a minor man page change. I just merged the PR.

**Comment by jarun** [comment]
Please review commit bf89040ace3105b6efde45ce3a078d5d594e5deb.

**Comment by jarun** [comment]
This is a really nice feature addition!

**Comment by leovilok** [comment]
To a nice file manager!

**Comment by jarun** [comment]
I have updated the wiki configuration and mouse/keyboard sections: https://github.com/jarun/nnn/wiki/Usage#configuration

**Pull request merged**